qiita_db/analysis.py

```diff
@@ -808,81 +808,122 @@ def build_files(self,
             # make testing much harder as we will need to have analyses at
             # different stages and possible errors.
             samples = self.samples
+
             # figuring out if we are going to have duplicated samples, again
             # doing it here cause it's computational cheaper
-            all_ids = list(chain.from_iterable([
-                samps for _, samps in viewitems(samples)]))
-            rename_dup_samples = ((len(all_ids) != len(set(all_ids))) and
-                                  merge_duplicated_sample_ids)
+            # 1. merge samples per data_type, reference used and the command id
+            #    if not one of the tables is not 'Pick closed-reference OTUs'
+            #    it's safer to always rename. Note that grouped_samples is
+            #    basically how many biom tables we are going to create
+            rename_dup_samples = False
+            grouped_samples = {}
+            for k, v in viewitems(samples):
+                a = qdb.artifact.Artifact(k)
+                p = a.processing_parameters
+                c = p.command
+                if c.name != 'Pick closed-reference OTUs':
+                    rename_dup_samples = True
+                    break
+                l = "%s.%d.%d" % (a.data_type, p.values['reference'], c.id)
+                if l not in grouped_samples:
+                    grouped_samples[l] = []
+                grouped_samples[l].append((k, v))
+            # 2. if rename_dup_samples is still False, make sure that we don't
+            #    need to rename samples by checking that there are not
+            #    duplicated samples per group
+            if not rename_dup_samples:
+                for k, v in viewitems(grouped_samples):
+                    # this element only has one table, continue
+                    if len(v) == 1:
+                        continue
+                    dup_samples = set()
+                    for k, v in v:
```

Comment:
Re-defining `v`, can we use better variable namaes? It's hard to follow the code as it is.